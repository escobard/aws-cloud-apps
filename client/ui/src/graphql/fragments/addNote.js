import gql from "graphql-tag";

export const ADD_NOTE_MUTATION = gql`
  mutation CreateNote($note: NoteInput) {
    createNote(note: $note) {
      id
      note
      subject
    }
  }
`