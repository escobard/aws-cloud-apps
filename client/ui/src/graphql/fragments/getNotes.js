import gql from "graphql-tag";

export const getNotes = gql`
  query GetNotes {
    getNotes {
    id
      date
      note
      subject
    }
  }
`