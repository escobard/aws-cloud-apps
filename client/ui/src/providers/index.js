'use client';

import NotesProvider, { NotesContext } from "./notes";

export { NotesProvider, NotesContext };
