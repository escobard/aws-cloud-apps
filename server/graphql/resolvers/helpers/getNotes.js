import { cacheHydrate } from "../../utils/cacheHydrate.js";

const getNotes = async (parent, knex) => {
    try{
        let notes;
        // checks to see if cache has data before calling DB
        /// if cache has no data, fetch data from database
        if (cache.keys().length === 0){
            notes = await cacheHydrate('notes.notes', '*', knex)
        }
        /// if cache has data, fetch data from cache instead of db
        else {
            notes = []
            cache.keys().map(id => {
                let note = cache.get(id)
                notes.push(note)
            });
        }
        // reverse order to return latest notes first

        return notes.reverse();
    }
    catch(err){
        return 'getNotes error:' + err;
    }
};

export default getNotes;
