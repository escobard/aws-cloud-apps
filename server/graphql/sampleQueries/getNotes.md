# getNotes query

Use this query to retrieve all available notes from the database / embedded API cache:

```
query GetNotes {
  getNotes {
  id
    date
    note
    subject
  }
}
```

## Expected response

This query should return an array{} of notes:

```
{
  "data": {
    "getNotes": [
      {
        "id": 1,
        "date": "2020-01-01T00:00:00.000Z",
        "note": "This is a note",
        "subject": "This is a subject"
      },
      {
        "id": 2,
        "date": "2020-01-01T00:00:00.000Z",
        "note": "This is a note",
        "subject": "This is a subject"
      },
      {
        "id": 3,
        "date": "2020-01-01T00:00:00.000Z",
        "note": "This is a note",
        "subject": "This is a subject"
      }
    ]
  }
}
```