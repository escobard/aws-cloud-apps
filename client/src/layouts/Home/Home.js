import React, { useContext } from "react";

import { NotesContext } from "../../providers";

import { Notes } from "../../components";

import "../../styles/global.scss";

import Form from "../../components/Form";
import { addNoteFields, notes as noteCatalog } from "../../constants";

const Home = () => {
  const { renderModal, notes, updateNotes, loading } = useContext(NotesContext);

  const renderNotes = (id, data) => {
    if (!data){
      return <Notes data={noteCatalog.loading} />;
    } else{
      const hasData = Array.isArray(data) && data.length > 0 && data[0].id;
      const hasApiError = !Array.isArray(data) && data.length > 0;

      if (hasData) {
        return data.map((note) => {
          return <Notes key={note.id} data={note} />;
        });
      }

      if (hasApiError) {
        noteCatalog.apiError.note = data;
        return <Notes data={noteCatalog.apiError} />;
      }

      return <Notes data={noteCatalog.noNotes} />;
    }

  };

  return (
    <>
      <main className="home">
        {renderModal({
          title: "Add note",
          content: <Form submit={updateNotes} fields={addNoteFields} />,
        })}
        {renderNotes("notes", notes)}
      </main>
    </>
  );
};

export default Home;
