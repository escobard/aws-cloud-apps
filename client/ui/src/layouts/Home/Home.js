import React, { useContext } from "react";

import { NotesContext } from "../../providers";

import { Notes } from "../../components";

import "../../styles/global.scss";

import Form from "../../components/Form";
import { addNoteFields, notes as noteCatalog } from "../../constants";

// TODO - move this entire file under app/page
const Home = () => {
  const { renderModal, notes, updateNotes, loading } = useContext(NotesContext);

  const renderNotes = (id, data) => {
    if (!data){
      return <Notes data={noteCatalog.loading} />;
    } else{
      const hasData = Array.isArray(data) && data.length > 0;
      const hasApiError = !Array.isArray(data) && data.length > 0;

      if (hasData) {
        return data.map((note, i) => {
          return <Notes key={note.subject + i} data={note} />;
        });
      }

      if (hasApiError) {
        noteCatalog.apiError.note = data;
        return <Notes data={noteCatalog.apiError} />;
      }

      return <Notes data={noteCatalog.noNotes} />;
    }

  };

  return (
    <>
      <main className="home">
        {renderModal({
          title: "Add note",
          content: <Form submit={updateNotes} fields={addNoteFields} />,
        })}
        {renderNotes("notes", notes)}
      </main>
    </>
  );
};

export default Home;
