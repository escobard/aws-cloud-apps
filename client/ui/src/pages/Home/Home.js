import React, { useContext } from "react";

import { NotesContext } from "../../providers";

import { Notes } from "../../components";

import "../../styles/global.scss";

import Form from "./Form";
import { addNoteFields, notes as noteCatalog } from "../../constants";

const Home = () => {
  const { renderModal, notes, updateNotes } = useContext(NotesContext);
  console.log(notes)
  const renderNotes = (id, data) => {
    const hasData = Array.isArray(data) && data.length > 0;
    const hasNoData = !Array.isArray(data) && data.length > 0;

    if (hasData) {
      return data.map((note, i) => {
        return <Notes key={note.subject + i} data={note} />;
      });
    }

    if (hasNoData) {
      noteCatalog.apiError.note = data;
      return <Notes data={noteCatalog.apiError} />;
    }

    return <Notes data={noteCatalog.noNotes} />;
  };

  return (
    <>
      <main className="home">
        {renderModal({
          title: "Add note",
          content: <Form submit={updateNotes} fields={addNoteFields} />,
        })}
        {notes && renderNotes("notes", notes)}
      </main>
    </>
  );
};

export default Home;
