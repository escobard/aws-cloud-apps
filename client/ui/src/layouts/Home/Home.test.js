import { NotesProvider } from "@/providers";

import Home from "./index";
import {GET_NOTES_QUERY} from "@/graphql/fragments/getNotes";

describe(">> render()", () => {

  const mocks = [
    {
      request: {
        query: GET_NOTES_QUERY
      },
      result: {
        data: {
          getNotes: []
        }
      }
    }
  ]

  it(">> snapshot with data is up to date", async () => {
    const { container } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <NotesProvider>
            <Home />
        </NotesProvider>
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();
  });
});
