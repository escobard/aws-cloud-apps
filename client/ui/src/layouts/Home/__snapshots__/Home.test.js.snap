// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>> render() >> generates snapshot 1`] = `
<div>
  <main
    class="home"
  >
    <div
      class="ui equal width grid note"
    >
      <div
        class="one wide computer three wide mobile two wide tablet column icon"
      >
        <i
          aria-label="note icon"
          class="sticky note big circular icon error"
        />
      </div>
      <div
        class="twelve wide computer nine wide mobile twelve wide tablet column content"
      >
        <h3>
          Loading...
        </h3>
        <p>
          Notes are loading...
        </p>
      </div>
    </div>
  </main>
</div>
`;
